src/mongoose_router.erl

```diff
@@ -0,0 +1,62 @@
+-module(mongoose_router).
+
+-define(TABLE, ?MODULE).
+
+-export([start/0, default_routing_modules/0]).
+
+-export([get_all_domains/0, lookup_route/1, is_registered_route/1,
+         register_route/2, unregister_route/1]).
+
+-spec get_all_domains() -> [jid:lserver()].
+get_all_domains() ->
+    ets:select(?TABLE, [{{'$1', '_'}, [], ['$1']}]).
+
+-spec lookup_route(jid:lserver()) -> no_route | mongoose_packet_handler:t().
+lookup_route(LDomain) ->
+    case ets:lookup(?TABLE, LDomain) of
+        [] -> no_route;
+        [{_, Handler}] ->
+            Handler
+    end.
+
+-spec register_route(jid:server(), mongoose_packet_handler:t()) -> any().
+register_route(Domain, Handler) ->
+    case jid:nameprep(Domain) of
+        error ->
+            {error, invalid_domain, Domain};
+        LDomain ->
+            ets:insert(?TABLE, {LDomain, Handler}),
+            mongoose_hooks:register_subhost(LDomain, false)
+    end.
+
+-spec unregister_route(jid:server()) -> any().
+unregister_route(Domain) ->
+    case jid:nameprep(Domain) of
+        error ->
+            {error, invalid_domain, Domain};
+        LDomain ->
+            ets:delete(?TABLE, LDomain),
+            mongoose_hooks:unregister_subhost(LDomain)
+    end.
+
+-spec is_registered_route(jid:server()) -> boolean().
+is_registered_route(Domain) ->
+    case jid:nameprep(Domain) of
```

Comment:
we don't need jid:nameprep here.
is_register_route should table jid:lserver() type.
(we already do jid:nameprep before calling the function).

We could require Domain also be prepped in register/unregister, it would be a bit more consistent, but I think we don't call register/unregister that often to care.